tests/phpunit/tests/admin/wpUpgrader.php

```diff
@@ -0,0 +1,1430 @@
+					),
+					'custom_path/subdir2'              => array(
+						'files' => array(
+							'subfile3.php' => array( 'name' => 'subfile3.php' ),
+							'subfile4.php' => array( 'name' => 'subfile4.php' ),
+							'subsubdir2'   => array(
+								'files' => array(
+									'subsubfile3.php' => array(
+										'name' => 'subsubfile3.php',
+									),
+									'subsubfile4.php' => array(
+										'name' => 'subsubfile4.php',
+									),
+								),
+							),
+						),
+					),
+					'custom_path/subdir2/subfile3.php' => array(
+						'name' => 'subfile3.php',
+					),
+					'custom_path/subdir2/subfile4.php' => array(
+						'name' => 'subfile4.php',
+					),
+					'custom_path/subdir2/subsubdir2'   => array(
+						'files' => array(
+							'subsubfile3.php' => array(
+								'name' => 'subsubfile3.php',
+							),
+							'subsubfile4.php' => array(
+								'name' => 'subsubfile4.php',
+							),
+						),
+					),
+					'custom_path/subdir2/subsubdir2/subsubfile3.php' => array(
+						'name' => 'subsubfile3.php',
+					),
+					'custom_path/subdir2/subsubdir2/subsubfile4.php' => array(
+						'name' => 'subsubfile4.php',
+					),
+				),
+				'nested_files' => array(
+					'subdir1' => array(
+						'files' => array(
+							'subfile1.php' => array( 'name' => 'subfile1.php' ),
+							'subfile2.php' => array( 'name' => 'subfile2.php' ),
+							'subsubdir1'   => array(
+								'files' => array(
+									'subsubfile1.php' => array(
+										'name' => 'subsubfile1.php',
+									),
+									'subsubfile2.php' => array(
+										'name' => 'subsubfile2.php',
+									),
+								),
+							),
+						),
+					),
+					'subdir2' => array(
+						'files' => array(
+							'subfile3.php' => array( 'name' => 'subfile3.php' ),
+							'subfile4.php' => array( 'name' => 'subfile4.php' ),
+							'subsubdir2'   => array(
+								'files' => array(
+									'subsubfile3.php' => array(
+										'name' => 'subsubfile3.php',
+									),
+									'subsubfile4.php' => array(
+										'name' => 'subsubfile4.php',
+									),
+								),
+							),
+						),
+					),
+				),
+				'path'         => 'custom_path/',
+			),
+		);
+	}
+
+	/**
+	 * Tests that `WP_Upgrader::clear_destination()` returns early with `true`
+	 * when the destination does not exist.
+	 *
+	 * @ticket 54245
+	 *
+	 * @covers WP_Upgrader::clear_destination
+	 */
+	public function test_clear_destination_should_return_early_when_the_destination_does_not_exist() {
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'is_writable' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'chmod' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'delete' );
+
+		$destination = DIR_TESTDATA . '/upgrade/';
+
+		$this->wp_filesystem_mock
+				->expects( $this->once() )
+				->method( 'dirlist' )
+				->with( $destination )
+				->willReturn( false );
+
+		$this->assertTrue( $this->instance->clear_destination( $destination ) );
+	}
```

Comment:
Not so much a remark about the test, but about the code in the function under test:
https://github.com/WordPress/wordpress-develop/blob/8db6a6e26af36004bd03ec7d9dd01e4427e6f0ad/src/wp-admin/includes/class-wp-upgrader.php#L391

I think this needs to also check for an empty array (directory exists, but contains no files).

The above test should then be updated to use a data provider to test both the `willReturn( false )` as well as the `willReturn( array() )` cases.

Reply:
Upon review of the source above and of `WP_Filesystem_Direct/FTPext/ftpsockets`, I believe you're right about needing to check for `array()`.

This may need a separate ticket for discussion/history and can be done after this PR is merged (or before, but I'm hoping this PR will be merged in the coming weeks).